File: prime-factors-java/src/main/java/PrimeFactors.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class PrimeFactors {

    /**
     *  1) 첫 번째 코드 
     *  ArrayList<Integer> factors = new ArrayList<>();
        if (n > 1) {
            factors.add(n);
            return factors;
        }
        
        return factors;


        ArrayList<Integer> factors = new ArrayList<>();
        if (n > 1) {

            //
            // 2) 두 번째 수정
            // if (n % 2 == 0) {
            //     factors.add(2);
            //     n /= 2;
            // }
            // 
            while (n % 2 == 0) {
                factors.add(2);
                n /= 2;
            }

        }
        if (n > 1) {
            factors.add(n);
        }
        return factors;

     */
    public static List<Integer> factorsOf(int n) {

        // if (n <= 1) {
        //     return Collections.emptyList();
        // }

        ArrayList<Integer> factors = new ArrayList<>();

        for (int divisor = 2; n > 1; divisor++) {
            for (; n % divisor == 0; n /= divisor) {
                factors.add(divisor);
            }
        }
        return factors;



        // throw new UnsupportedOperationException("Prime factorization is not implemented yet.");
    }
}

File: fn-clojure/prime-factos-clojure/src/prime_factors.clj
(ns prime-factors)

;; (defn factors-of [n]
;;   (if (<= n 1)
;;     []
;;     (throw (UnsupportedOperationException.
;;              "Prime factorization is not implemented yet."))))

;; 1번째 (defn prime-factors-of [n] [])

;; 2번째 
;; (defn prime-factors-of [n] 
;;   (if (> n 1) [2] []))

;; 3번째
;; (defn prime-factors-of [n] 
;;   (if (> n 1) [n] []))

;; 4번째 
;; (defn prime-factors-of [n] 
;;   (if (> n 1) 
;;     (if (zero? (rem n 2))
;; ;; cons 함수 - prime-factors-of에서 반환된 리스트의 시작 부분에 2를 추가
;;       (cons 2 (prime-factors-of (quot n 2)))
;;       [n])
;;   [])
;; )

;; 
(defn prime-factors-of [n] 
;; loop - 즉석에서 새로운 익명 함수 생성
  (loop [n n 
        divisor 2
        factors []]
    (if (> n 1) 
      (if (zero? (rem n divisor))
        ;; conj 함수 - factors에 값을 추가
        (recur (quot n divisor) divisor (conj factors divisor))
        (recur n (inc divisor) factors))
      factors
    [])
  )
)

